src/lib.rs

```diff
@@ -424,30 +424,50 @@ impl<R: gimli::Reader> FullContext<R> {
 
 type Error = gimli::Error;
 
-fn str_attr<'abbrev, 'unit, R: gimli::Reader>(
+fn name_attr<'abbrev, 'unit, R: gimli::Reader>(
     entry: &gimli::DebuggingInformationEntry<'abbrev, 'unit, R, R::Offset>,
     dw_unit: &gimli::CompilationUnitHeader<R, R::Offset>,
     abbrevs: &gimli::Abbreviations,
     sections: &DebugSections<R>,
-    name: gimli::DwAt,
 ) -> Result<Option<R>, Error> {
-    Ok(match entry.attr(name)? {
-        Some(x) => x.string_value(&sections.debug_str),
-        None => {
-            match entry.attr_value(gimli::DW_AT_abstract_origin)? {
-                Some(gimli::AttributeValue::UnitRef(offset)) => {
-                    let mut tcursor = dw_unit.entries_at_offset(abbrevs, offset)?;
-                    match tcursor.next_dfs()? {
-                        // FIXME: evil dwarf can send us into an infinite loop here
-                        Some((_, entry)) => str_attr(entry, dw_unit, abbrevs, sections, name)?,
-                        None => None,
-                    }
-                }
-                None => None,
-                x => panic!("wat {:?}", x),
+    if let Some(attr) = entry.attr(gimli::DW_AT_linkage_name)? {
+        if let Some(val) = attr.string_value(&sections.debug_str) {
+            return Ok(Some(val));
+        }
+    }
+    if let Some(attr) = entry.attr(gimli::DW_AT_MIPS_linkage_name)? {
+        if let Some(val) = attr.string_value(&sections.debug_str) {
+            return Ok(Some(val));
+        }
+    }
+    if let Some(attr) = entry.attr(gimli::DW_AT_name)? {
+        if let Some(val) = attr.string_value(&sections.debug_str) {
+            return Ok(Some(val));
+        }
+    }
+    match entry.attr_value(gimli::DW_AT_abstract_origin)? {
+        Some(gimli::AttributeValue::UnitRef(offset)) => {
+            let mut entries = dw_unit.entries_at_offset(abbrevs, offset)?;
+            if let Some((_, entry)) = entries.next_dfs()? {
+                // FIXME: evil dwarf can send us into an infinite loop here
```

Comment:
Does the spec even allow chains of redirections? If not, do compilers adhere to that? Then we could solve this with a simple flag.

Reply:
Technically I think it's not allowed, but gdb has [a comment](https://sourceware.org/git/gitweb.cgi?p=binutils-gdb.git;a=blob;f=gdb/dwarf2read.c;h=b6bd2917e35f378e983cf959d1273399ab8ccf65;hb=HEAD#l12326) saying that gcc generates it. I think a simple recursion count with a low limit will be enough.